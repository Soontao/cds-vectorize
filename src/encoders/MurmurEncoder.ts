import { murmurhash3 } from "../utils/murmurhash3.js";
import Encoder from "./Encoder.js";
import { Vector } from "./type.js";
import { isNull, isNullVector, nullVector } from "./util.js";

/**
 * in most case, use this for semantic code
 *
 * @ai
 * @human
 */
export class MurmurEncoder implements Encoder<string> {
  #seed: number;

  constructor(seed: number = 42) {
    this.#seed = seed;
  }

  encode(value: string): Vector {
    if (isNull(value)) {
      return nullVector(this.length);
    }
    return [murmurhash3(String(value), this.#seed)];
  }

  decode(_vec: Vector): string {
    if (isNullVector(_vec)) {
      return null as any;
    }
    throw new Error("Decoding is not implemented for MurmurEncoder.");
  }

  features(name: string): Array<string> {
    return [name];
  }

  get length(): number {
    return 1;
  }
}

export default MurmurEncoder;
