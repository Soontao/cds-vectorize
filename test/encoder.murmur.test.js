import { MurmurEncoder } from "../src/encoders/MurmurEncoder.js";

describe("MurmurEncoder Test Suite", () => {
  it("should support int test", () => {
    expect(new MurmurEncoder(42).encode("Hello Vector!")).toMatchSnapshot();
    expect(new MurmurEncoder(500).encode("Hello Vector!")).toMatchSnapshot();
  });
  it("should have a length of 1", () => {
    const encoder = new MurmurEncoder();
    expect(encoder.length).toEqual(1);
  });

  it("should return feature array with the provided name", () => {
    const encoder = new MurmurEncoder();
    const featureName = "test_feature";

    const features = encoder.features(featureName);

    expect(features).toEqual([featureName]);
  });

  it("should throw an error when decoding is attempted", () => {
    const encoder = new MurmurEncoder();
    const encodedVector = [123];

    expect(() => encoder.decode(encodedVector)).toThrow("Decoding is not implemented for MurmurEncoder.");
  });
});
