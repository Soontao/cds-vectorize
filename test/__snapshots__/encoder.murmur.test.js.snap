// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MurmurEncoder Test Suite should support int test 1`] = `
[
  4212377875,
]
`;

exports[`MurmurEncoder Test Suite should support int test 2`] = `
[
  2026014530,
]
`;

exports[`MurmurEncoder Test Suite should support int test 3`] = `
[
  -65536,
]
`;

exports[`MurmurEncoder Test Suite should support int test 4`] = `
[
  -65536,
]
`;
